zkevm-circuits/src/evm_circuit/execution/error_return_data_oo_bound.rs

```diff
@@ -0,0 +1,358 @@
+    is_end_within_range: IsZeroGadget<F>,
+    // when `end` not overflow, check if it exceeds return data size.
+    is_end_exceed_length: LtGadget<F, N_BYTES_U64>,
+    rw_counter_end_of_reversion: Cell<F>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ErrorReturnDataOutOfBoundGadget<F> {
+    const NAME: &'static str = "ErrorReturnDataOutOfBoundGadget";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorReturnDataOutOfBound;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        let memory_offset = cb.query_cell();
+        let data_offset = cb.query_word_rlc();
+        let length = cb.query_word_rlc();
+        let end = cb.query_word_rlc();
+        let rw_counter_end_of_reversion = cb.query_cell();
+
+        let return_data_length = cb.query_cell();
+
+        cb.require_equal(
+            "opcode is RETURNDATACOPY",
+            opcode.expr(),
+            OpcodeId::RETURNDATACOPY.expr(),
+        );
+
+        // Pop memory_offset, offset, length from stack
+        cb.stack_pop(memory_offset.expr());
+        cb.stack_pop(data_offset.expr());
+        cb.stack_pop(length.expr());
+
+        // read last callee return data length
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::LastCalleeReturnDataLength,
+            return_data_length.expr(),
+        );
+
+        // check `data_offset` u64 overflow
+        let data_offset_larger_u64 = from_bytes::expr(&data_offset.cells[8..]);
+        let is_data_offset_within_range = IsZeroGadget::construct(cb, data_offset_larger_u64);
+
+        // when data set is not overflow, check sum and `end` if not overflow.
+        let (sum, is_end_within_range) = cb.condition(is_data_offset_within_range.expr(), |cb| {
+            let sum_gadget =
+                AddWordsGadget::construct(cb, [data_offset.clone(), length], end.clone());
+            let end_larger_u64 = from_bytes::expr(&end.cells[8..]);
+            (sum_gadget, IsZeroGadget::construct(cb, end_larger_u64))
+        });
+
+        // when `end` not u64 overflow, check if it exceeds return data length
+        let is_end_exceed_length = cb.condition(
+            is_data_offset_within_range.expr() * is_end_within_range.expr(),
+            |cb| LtGadget::construct(cb, return_data_length.expr(), end.expr()),
+        );
+
+        // only one of (data offset overflow, end overflow, end_exceed_length) must be
+        // true. because when data offset overflow, don't go to check other two,
+        // when `end` overflow, don't check the last. which is accordance to evm
+        // execution.
+        cb.require_equal(
+            "name",
+            not::expr(is_data_offset_within_range.expr())
+                + not::expr(is_end_within_range.expr())
+                + is_end_exceed_length.expr(),
+            1.expr(),
+        );
```

Comment:
```suggestion
        cb.require_in_set(
            "Any of [offset_out_of_range, end_out_of_range, end_exceed_length]",
            not::expr(is_data_offset_within_range.expr())
                + not::expr(is_end_within_range.expr())
                + is_end_exceed_length.expr(),
            vec![1.expr(), 2.expr(), 3.expr()],
        );
```

Reply:
updated in [f3bc7ac](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1017/commits/f3bc7ac6898a06c6ef724f12875b74410bb0a81d) thanks.